// pagedjs/lib/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Chunker", {
  enumerable: true,
  get: function get() {
    return _chunker["default"];
  }
});
Object.defineProperty(exports, "Handler", {
  enumerable: true,
  get: function get() {
    return _handler["default"];
  }
});
Object.defineProperty(exports, "Polisher", {
  enumerable: true,
  get: function get() {
    return _polisher["default"];
  }
});
Object.defineProperty(exports, "Previewer", {
  enumerable: true,
  get: function get() {
    return _previewer["default"];
  }
});
Object.defineProperty(exports, "initializeHandlers", {
  enumerable: true,
  get: function get() {
    return _handlers.initializeHandlers;
  }
});
Object.defineProperty(exports, "registerHandlers", {
  enumerable: true,
  get: function get() {
    return _handlers.registerHandlers;
  }
});
Object.defineProperty(exports, "registeredHandlers", {
  enumerable: true,
  get: function get() {
    return _handlers.registeredHandlers;
  }
});
var _chunker = _interopRequireDefault(require("./chunker/chunker.cjs"));
var _polisher = _interopRequireDefault(require("./polisher/polisher.cjs"));
var _previewer = _interopRequireDefault(require("./polyfill/previewer.cjs"));
var _handler = _interopRequireDefault(require("./modules/handler.cjs"));
var _handlers = require("./utils/handlers.cjs");
